### PR TITLE
Unset bindings in psci

### PR DESCRIPTION
**Pull request by stevejb71**
This is intended to address #1030 

Seems to work and is simple, so hopefully not too hard to review. 

This is different to ghci, where the last result would be 15:

```
> import Prelude 
> let x = 10
> let f a = x + a
> f 5
15

> let x = 20
> f 5
25
```

### CODE REVIEW

File: psci/PSCi/Types.hs [hunk 1]
```diff
@@ -121,12 +121,17 @@ updateModules modules st =
...
 -- |
 -- Updates the state to have more let bindings.
+-- Replaces ValueDeclarations if they have the same identifier.
 --
 updateLets :: [P.Declaration] -> PSCiState -> PSCiState
-updateLets ds st = st { psciLetBindings = psciLetBindings st ++ ds }
-
+updateLets ds st = st { psciLetBindings = keptBindings ++ ds }
+  where keptBindings = filter shouldKeep (psciLetBindings st)
+        shouldKeep decl = valueDeclarationIdent decl `notElem` map valueDeclarationIdent ds
```
Comment (paf31) [inline psci/PSCi/Types.hs]: Why not use a `Map`?

Comment (stevejb71) [inline psci/PSCi/Types.hs]: Not sure what to key against. If it were just value declarations I'd use ```Ident```, but ```psciLetBindings```can include data declarations or types as well (it looks misnamed to me).

**Comment by paf31** [comment]
I'd consider using `Either` or creating your own sum type.

**Comment by stevejb71** [comment]
Here's a go at using a Map to join the declarations into ```PSCiState```
Producing keys for all of the ```P.Declaration``` is a little messy and I don't think all of them are used in PSCi (e.g. ```BindingGroupDeclaration```).
So I've tried just putting ValueDeclarations into a Map and having a separate field in ```PSCiState``` for non-```ValueDeclaration```s.

My original commit seems simpler though - what is the benefit of using a Map?

**Comment by LiamGoodacre** [comment]
I'm not happy with your example/suggested behaviour:

~~~purescript
> import Prelude 
> let x = 10
> let f a = x + a
> f 5
15

> let x = 20
> f 5
25
~~~

I feel that `f` should still give `15` - it was defined in terms of `x_0` not `x_1`.  Consider the following potentially confusing result:

~~~purescript
> let a = 10
> let b = a * 10
> let f x = [x, a, b]
> f 0
[0, 10, 100]

> let a = 20
> f 0
[0, 20, 100]
~~~

Here by replacing `a` you've changed `f` but not `b`, even though they were both defined in terms of a previous `a`.

If you want to take advantage of this behaviour you will have to be careful never to be point-free.  If you mix them up then you'll get strange results and have to re-`let` everything anyway.

What are your thoughts on this? @stevejb71 @paf31

**Comment by AppShipIt** [comment]
> What are your thoughts on this?

I agree this would be unpleasantly surprising. I expect new bindings to shadow new uses not pervious ones.

**Comment by paf31** [comment]
I think maybe a warning is a good enough way to handle this situation? Changing the meaning of the inputs to PSCi from a list of declarations in a module to something which tracks dependencies is going to be quite a lot more work.

File: psci/PSCi/Types.hs [hunk 2]
```diff
@@ -118,15 +125,17 @@ updateModules modules st =
...
 --
-updateLets :: [P.Declaration] -> PSCiState -> PSCiState
-updateLets ds st = st { psciLetBindings = psciLetBindings st ++ ds }
-
+updateDeclarationBindings :: [P.Declaration] -> PSCiState -> PSCiState
+updateDeclarationBindings ds st = st { psciValueDeclarations = valueDeclarations `Map.union` psciValueDeclarations st, psciDeclarations = psciDeclarations st ++ otherDeclarations }
+  where (valueDeclarations, otherDeclarations) = B.first (Map.fromList . map (\d -> (identFor d, d))) . L.partition P.isValueDecl $ ds
+        identFor (P.ValueDeclaration i _ _ _) = i
+        identFor (P.PositionedDeclaration _ _ x) = identFor x
+        identFor _ = error "this should not have happened"
```
Comment (paf31) [inline psci/PSCi/Types.hs]: Could you provide a better error and use `internalError` here please?

**Comment by paf31** [comment]
One other downside:

```
> import Prelude 
> let x = 10
> let f a = x + a
> f 5
15

> let x = "explode"
```

will now give an error.

**Comment by hdgarrood** [comment]
Especially given this most recent example. I'm tempted to say we should at least investigate how hard it would be to avoid this. Maybe tracking dependencies isn't necessary, maybe we can add numeric suffixes to reused identifiers behind the scenes or something?

**Comment by AppShipIt** [comment]
When I watched the iHaskell talk he described GHCi as if each instruction introduced a new `do` block. Would that be an effective solution? New names would shadow old ones but old ones would continue to be closed over.

**Comment by paf31** [comment]
@AppShipIt I thought about that, but it's not clear how or when to put things in a `let`.

Suppose you define `a` and then `b` and then `c`. `b` and `c` use `a`. Then you redefine `a`. Should the existing `a` be pushed into `let` clauses in `b` and `c` and the new `a` be defined at the top level? Until something is redefined it needs to be available for use at the top level.

An alternative could be to always push dependencies into `let` expressions, duplicating them as many times as is needed. It might result in code explosion though. If we did it the naive way, you'd end up with things like:

```purescript
let a = ...
     b = let a = ...
           in f a
     c = let a = ...
               b = let a = ...
                       in f a
           in g a b
in h a b c
```

File: psci/PSCi/Types.hs [hunk 3]
```diff
@@ -118,15 +125,17 @@ updateModules modules st =
   st { _psciLoadedModules = Map.union (go modules) (_psciLoadedModules st) }
   where
   go = Map.fromListWith (++) . map (second (:[]))
-
```
Comment (AppShipIt) [inline psci/PSCi/Types.hs]: I think this deleted line was accidental.

File: psci/PSCi/Types.hs (hunk 2 above)
Comment (AppShipIt) [inline psci/PSCi/Types.hs]: https://github.com/purescript/purescript/blob/279527f200ca5d6c1fe890e09aa86fe9a847c78b/src/Language/PureScript/Sugar/BindingGroups.hs#L160

**Comment by AppShipIt** [comment]
IIUC @stevejb71 changed `let` bindings to a map that overwrites duplicate identifiers? My suggestion seems <q>non-trivial</q> by comparison. I feel the current behavior is preferable until addressed architecturally.

**Comment by hdgarrood** [comment]
> Should the existing a be pushed into let clauses in b and c and the new a be defined at the top level?

That seems reasonable to me, yeah. Is the problem with this approach just that it would be complex?

Alternatively, going for the naive route, maybe we can put up with the code explosion for now, and when we get around to it, implement common subexpression elimination in the compiler?

**Comment by paf31** [comment]
Closing due to inactivity, but we can reopen if the issues above can be resolved.

**Pull request closed without merging**